src/pharmpy/internals/fs/tmp.py

```diff
@@ -1,83 +1,92 @@
-                        try:
-                            os.chflags(path, 0)
-                        except AttributeError:
-                            pass
-                        os.chmod(path, 0o700)
+    @classmethod
+    def _rmtree(cls, name, ignore_errors=False):
+        def onerror(func, path, exc_info):
+            if issubclass(exc_info[0], PermissionError):
 
+                def resetperms(path):
                     try:
-                        if path != name:
-                            resetperms(os.path.dirname(path))
-                        resetperms(path)
-
-                        try:
-                            os.unlink(path)
-                        # PermissionError is raised on FreeBSD for directories
-                        except (IsADirectoryError, PermissionError):
-                            time.sleep(0.1)
-                            cls._rmtree(path)
-                    except FileNotFoundError:
+                        _os.chflags(path, 0)
+                    except AttributeError:
                         pass
-                elif issubclass(exc_info[0], FileNotFoundError):
+                    _os.chmod(path, 0o700)
+
+                try:
+                    if path != name:
+                        resetperms(_os.path.dirname(path))
+                    resetperms(path)
+
+                    try:
+                        _os.unlink(path)
+                    # PermissionError is raised on FreeBSD for directories
+                    except (IsADirectoryError, PermissionError):
+                        cls._rmtree(path, ignore_errors=ignore_errors)
+                except FileNotFoundError:
                     pass
-                else:
+            elif issubclass(exc_info[0], FileNotFoundError):
+                pass
+            else:
+                if not ignore_errors:
                     raise
 
-            shutil.rmtree(name, onerror=onerror)
-
-        @classmethod
-        def _cleanup(cls, name, warn_message):
-            cls._rmtree(name)
-            warnings.warn(warn_message, ResourceWarning)
+        _shutil.rmtree(name, onerror=onerror)
 
-        def __repr__(self):
-            return "<{} {!r}>".format(self.__class__.__name__, self.name)
+    @classmethod
+    def _cleanup(cls, name, warn_message, ignore_errors=False):
+        cls._rmtree(name, ignore_errors=ignore_errors)
+        _warnings.warn(warn_message, ResourceWarning)
 
-        def __enter__(self):
-            return self.name
+    def __repr__(self):
+        return "<{} {!r}>".format(self.__class__.__name__, self.name)
 
-        def __exit__(self, exc, value, tb):
-            self.cleanup()
+    def __enter__(self):
+        return self.name
 
-        def cleanup(self):
-            if self._finalizer.detach():
-                self._rmtree(self.name)
+    def __exit__(self, exc, value, tb):
+        self.cleanup()
 
-else:
-    from tempfile import TemporaryDirectory  # noqa
```

Comment:
Using the same implementation on Windows, Linux, and MacOS makes it simpler to do what we want without having to write too much extra code. Otherwise we would have to have some additional code conditional on `os.name` to figure out if we need to pass the `ignore_cleanup_errors` flag. Here we can just use `TemporaryDirectory` as if it was coming from Python 3.10. There has not been any change to the source since then, so the implementation is exactly the same in Python 3.11. Moreover, I guess that if newer versions introduced important patches we would likely need to add those to our copy anyway. Once support for Python 3.8 and 3.9 is dropped, we can just discard the whole thing.